server/src/main/java/org/opensearch/threadpool/ThreadPool.java

```diff
@@ -222,10 +228,99 @@ public Collection<ExecutorBuilder> builders() {
+                    o.setCorePoolSize(size);
+                    o.setMaximumPoolSize(size);
+                } else {
+                    o.setMaximumPoolSize(size);
+                    o.setCorePoolSize(size);
+                }
+            }
+        }
+    }
+
+    public void setClusterSettings(ClusterSettings clusterSettings) {
+        this.clusterSettings = clusterSettings;
+        this.clusterSettings.addSettingsUpdateConsumer(CLUSTER_THREAD_POOL_SIZE_SETTING, this::setThreadPool, this::validateSetting);
+    }
+
+    private void validateSetting(Settings tpSettings) {
+        Map<String, Settings> tpGroups = tpSettings.getAsGroups();
+        for (Map.Entry<String, Settings> entry : tpGroups.entrySet()) {
+            String tpName = entry.getKey();
+            if (THREAD_POOL_TYPES.get(tpName) == null) {
+                throw new IllegalArgumentException("illegal thread_pool name : " + tpName);
+            }
+            Settings tpGroup = entry.getValue();
+            ExecutorHolder holder = executors.get(tpName);
+            assert holder.executor instanceof OpenSearchThreadPoolExecutor;
+            OpenSearchThreadPoolExecutor o = (OpenSearchThreadPoolExecutor) holder.executor;
+            if (holder.info.type == ThreadPoolType.SCALING) {
+                Set<String> so = new HashSet<>(Arrays.asList("max", "core"));
+                if (tpGroup.keySet().stream().allMatch(so::contains) == false) {
+                    throw new IllegalArgumentException(
+                        "illegal thread_pool config : " + tpGroup.keySet() + " should only have max and core"
+                    );
+                }
+                int max = tpGroup.getAsInt("max", o.getMaximumPoolSize());
+                int core = tpGroup.getAsInt("core", o.getCorePoolSize());
+                if (core < 1 || max < 1) {
+                    throw new IllegalArgumentException("illegal value for [cluster.thread_pool." + tpName + "], has to be positive value");
+                } else if (core > max) {
+                    throw new IllegalArgumentException("core threadpool size cannot be greater than max");
+                }
+            } else {
+                if (tpGroup.keySet().size() != 1) {
+                    throw new IllegalArgumentException("illegal thread_pool config : " + tpGroup.keySet());
+                } else if (tpGroup.keySet().contains("size") == false) {
+                    throw new IllegalArgumentException("illegal thread_pool config : " + tpGroup.keySet() + " should only have size");
+                } else {
+                    int size = tpGroup.getAsInt("size", o.getMaximumPoolSize());
+                    if (size < 1) {
+                        throw new IllegalArgumentException(
+                            "illegal value for [cluster.thread_pool." + tpName + "], has to be positive value"
+                        );
+                    }
+                }
+            }
+        }
```

Comment:
Why do we need to handle this differently?

Reply:
This is simplified now .